Send channel_update messages to direct peers on private channels

TheBlueMatt opened the pull request:
If we are a public node and have a private channel, our
counterparty needs to know the fees which we will charge to forward
payments to them. Without sending them a channel_update, they have
no way to learn that information, resulting in the channel being
effectively useless for outbound-from-us payments.

This commit fixes our lack of channel_update messages to private
channel counterparties, ensuring we always send them a
channel_update after the channel funding is confirmed.

Still need to fix existing tests/add new tests for the new behavior.

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1537,31 +1536,31 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
-						break Some(("Forwarding channel is not in a ready state.", 0x1000 | 20, Some(self.get_channel_update(chan).unwrap())));
+						break Some(("Forwarding channel is not in a ready state.", 0x1000 | 20, Some(self.get_channel_update_for_unicast(chan).unwrap())));
 					}
 					if *amt_to_forward < chan.get_counterparty_htlc_minimum_msat() { // amount_below_minimum
```
why send a channel update to the counterparty if the *counterparty* policy was violated?

TheBlueMatt replied:
Are you referring to the `amt_to_forward` value? It looks like the `htlc_minimum_msat` field in the generated channel update is `chan.get_counterparty_htlc_minimum_msat()`.

devrandom replied:
whoa, this got attached to the wrong line, sorry

TheBlueMatt replied:
No, this is being sent back to the HTLC sender. Practically its a no-op because the HTLC-sender shouldn't really store info about a private channel, but the spec says we have to include a channel_update for this error type. In reality, the bug is on the recipient's end, they shouldnt have included the channel hint in the invoice to begin with.

devrandom replied:
ah, got it now.  thank you.

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1537,31 +1536,31 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						break Some(("CLTV expiry is too far in the future", 21, None));
 					}
 					// In theory, we would be safe against unitentional channel-closure, if we only required a margin of LATENCY_GRACE_PERIOD_BLOCKS.
 					// But, to be safe against policy reception, we use a longuer delay.
```
nit: typos in last two lines

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2348,7 +2365,7 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 					ChannelUpdateStatus::EnabledStaged if !chan.is_live() => chan.set_channel_update_status(ChannelUpdateStatus::Disabled),
 					ChannelUpdateStatus::DisabledStaged if !chan.is_live() => {
-						if let Ok(update) = self.get_channel_update(&chan) {
+						if let Ok(update) = self.get_channel_update_for_broadcast(&chan) {
```
Don't we also want to send a unicast in this case if the channel is private?

TheBlueMatt replied:
Hmmm, I'd think it doesn't matter much - its not like our counterparty doesn't know that we're disconnected. I'm not sure what other nodes do, but we should correctly flag a channel as `!is_usable` in the `ChannelDetails` and thus not include it in invoices.

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -947,6 +947,11 @@ macro_rules! handle_chan_restoration_locked {
 						node_id: counterparty_node_id,
 						msg: announcement_sigs,
 					});
+				} else if $channel_entry.get().is_usable() {
```
This is gated by the `if let Some(msg) = $funding_locked {` above.  Does it make sense to always send a channel update on reconnect, just in case we updated our policy while the channel was disconnected?

TheBlueMatt replied:
Hmm, good point, I pushed the channel_update sending down into the macro callsites.

devrandom commented:
Tested - correctly sends a `channel_update` right after `funding_locked`

TheBlueMatt commented:
Fixed tests after rebasing on #954.

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3381,13 +3420,26 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 							node_id: counterparty_node_id.clone(),
 							msg,
 						});
+					} else if chan.get().is_usable() {
```
shouldn't this also check that the channel is private?

TheBlueMatt replied:
I don't think we need to, no - when reconnecting its good to make sure that our direct channel counterparty has an updated copy of our channel_update, whether its public or not.

ariard commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3953,6 +3963,12 @@ where
 								msg: announcement_sigs,
 							});
+						} else if channel.is_usable() {
+							log_trace!(self.logger, "Sending funding_locked WITHOUT announcement_signatures but with channel_update for {}", log_bytes!(channel.channel_id()));
```
Maybe add for "with channel_update for _private_", underscoring the privacy of the chan

ariard reviewed (commented):
Overall SGTM

TheBlueMatt commented on lightning/src/ln/channel.rs:
```diff
@@ -3996,7 +3996,7 @@ impl<Signer: Sign> Channel<Signer> {
 		}
 
-		if (self.channel_state & (ChannelState::PeerDisconnected as u32 | ChannelState::MonitorUpdateFailed as u32)) != 0 {
+		if (self.channel_state & (ChannelState::PeerDisconnected as u32)) != 0 {
```
Oops, indeed. Note that this is #954.

TheBlueMatt commented:
Rebased now that  #948 is merged, note that this still depends on # #948.

TheBlueMatt commented:
Rebased on #954 and latest upstream, squashing the commits in this PR and fixing conflicts, also rebased on #972 and added a new commit which extends the test from it and adds a new bugfix.

TheBlueMatt commented:
Rebased after all dependent PRs were merged.

TheBlueMatt commented:
Fixed bench + fuzz failures due to the new messages which they weren't expecting.

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3970,6 +4005,12 @@ where
 							});
+						} else if channel.is_usable() {
+							log_trace!(self.logger, "Sending funding_locked WITHOUT announcement_signatures but with private channel_update for our counterparty on channel {}", log_bytes!(channel.channel_id()));
+							pending_msg_events.push(events::MessageSendEvent::SendChannelUpdate {
+								node_id: channel.get_counterparty_node_id(),
+								msg: self.get_channel_update_for_unicast(channel).unwrap(),
+							});
```
Don't quite get why we don't push a `SendChannelUpdate` event when `announcement_sigs.is_some()`?

TheBlueMatt replied:
If we send a `channel_update` which we intend to have relayed, we have to send a `channel_announcement` first. If we're just sending it to our counterparty they shouldn't care, but if the channel is public we have to wait until announcement_signatures are exchanged so we can send a channel_announcement.

valentinewallace commented on lightning/src/util/events.rs:
```diff
@@ -389,6 +389,15 @@ pub enum MessageSendEvent {
+	/// Used to indicate that a channel_update should be sent to a single peer.
+	/// This is used, in contrast to [`Self::BroadcastChannelUpdate`], when the channel is a
+	/// private channel and we shouldn't be informing all of our peers of channel parameters.
+	SendChannelUpdate {
```
nit: prefer `UnicastChannelUpdate` slightly for consistency

TheBlueMatt replied:
Hmm, all our other message events are `Send`..., though? I could move it up to sit next to the rest of the send message variants?

valentinewallace replied:
Ah, true. Resolving

valentinewallace commented on lightning/src/util/events.rs:
```diff
@@ -389,6 +389,15 @@ pub enum MessageSendEvent {
 		msg: msgs::ChannelUpdate,
 	},
+	/// Used to indicate that a channel_update should be sent to a single peer.
+	/// This is used, in contrast to [`Self::BroadcastChannelUpdate`], when the channel is a
```
nit: rephrase to: In contrast to [Self::BroadcastChannelUpdate], this is used when the channel...

valentinewallace commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -3408,7 +3408,13 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 				}
-				try_chan_entry!(self, chan.get_mut().channel_update(&msg), channel_state, chan);
+				let were_node_one = self.get_our_node_id().serialize()[..] < chan.get().get_counterparty_node_id().serialize()[..];
+				let msg_from_node_one = msg.contents.flags & 1 == 0;
```
Might be nicer to make `ChannelFlags` into its own struct and have methods to check who's node_one, etc. Not necessary for this PR though

TheBlueMatt replied:
Yea, that's probably a good idea for the `channel_state` in `Channel` too, but we can do it later.

devrandom commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -2947,6 +2981,11 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						msg: announcement_sigs,
 					});
+				} else if chan.get().is_usable() {
+					channel_state.pending_msg_events.push(events::MessageSendEvent::SendChannelUpdate {
```
is it appropriate to send this message if we haven't seen the locking block?

TheBlueMatt replied:
Hmm, good point. The spec is mum on it (it only mentions that we have to wait until we *receive* a `funding_locked`) see https://github.com/lightningnetwork/lightning-rfc/blob/master/07-routing-gossip.md#requirements-3 so I think its fine. There isn't a ton of logical reason to wait, either, IMO, its just some fields the counterparty stores locally, so who cares when its received.

devrandom replied:
OK, something to keep in mind if we see interop issues in this area.

TheBlueMatt commented on lightning/src/ln/channelmanager.rs:
```diff
@@ -1587,9 +1586,29 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
+	/// [`MessageSendEvent::BroadcastChannelUpdate`] event.
+	///
+	/// May be called with channel_state already locked!
+	fn get_channel_update_for_broadcast(&self, chan: &Channel<Signer>) -> Result<msgs::ChannelUpdate, LightningError> {
```
Hmm, yea, happy to do something if you have a specific suggestion but I'm not sure what you really could do there. At least the `broadcast`/`unicast` naming makes it pretty clear IMO, especially when you're shoving it in a `Broadcast`... event

devrandom replied:
Could have two thin wrappers around `msg::ChannelUpdate` for unicast/broadcast, but I haven't looked in detail whether it really makes sense.

TheBlueMatt replied:
I dunno, its very explicit at the callsite already, making it more explicit doesn't seem like it'll solve anything IMO.

TheBlueMatt commented:
Squashed without diff, will merge after CI:
```
$ git diff-tree -U1 6cf22aa2 ba600db7
$
```
